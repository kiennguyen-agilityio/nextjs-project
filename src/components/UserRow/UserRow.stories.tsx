import type { Meta, StoryObj } from '@storybook/react';

// components
import UserRow from '@/components/UserRow';

const meta = {
  title: 'Components/UserInfo',
  component: UserRow,
} satisfies Meta<typeof UserRow>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    avatar: 'https://s.net.vn/Ur2Q',
    email: 'test@example.com',
    name: 'John Doe',
    role: 'Admin',
    joinDate: '2022-01-01',
  },
};
