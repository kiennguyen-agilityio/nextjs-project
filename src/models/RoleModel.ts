export type RoleModel = {
  id: string;
  name: string;
  description: string;
  createAt?: Date;
};
