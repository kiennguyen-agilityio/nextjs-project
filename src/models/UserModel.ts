export interface UserModel {
  id: string;
  name: string;
  email: string;
  userRole: string;
  avatar: string;
  joined: Date;
}
