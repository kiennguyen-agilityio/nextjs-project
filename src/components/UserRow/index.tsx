import Image from 'next/image';

// components
import { TripleDot } from '@/icons/TripleDot';

// models
import { UserModel } from '@/models/UserModel';
import { RoleModel } from '@/models/RoleModel';

// api
import { getRoleById } from '@/api/role';

const UserRow = async ({
  avatar,
  name,
  email,
  userRole,
  joined,
}: UserModel) => {
  const role: RoleModel = await getRoleById(userRole);

  return (
    <tr className="flex flex-col sm:flex-row items-start sm:items-center sm:gap-10 pr-8 md:pr-0 justify-between p-0 sm:p-4 border-b border-gray-200 hover:bg-gray-100 transition-colors duration-200 ease-in-out">
      <td className="flex items-center w-full mb-4 sm:mb-0">
        <Image
          src={avatar}
          alt={name}
          className="w-10 h-10 rounded-lg mr-4"
          width={40}
          height={40}
        />
        <div className="text-sm font-medium sm:truncate max-w-24 xl:max-w-64 md:max-w-25">
          {email}
        </div>
      </td>
      <td className="text-sm w-full truncate mb-4 sm:mb-0">{name}</td>
      <td className="text-sm w-full truncate mb-4 sm:mb-0">{role.name}</td>
      <td className="text-sm w-full truncate mb-4 sm:mb-0">
        {new Date(joined).toLocaleDateString()}
      </td>
      <td className="w-full sm:w-1/12">
        <button className="text-gray-500" data-testid="triple-dot">
          <TripleDot />
        </button>
      </td>
    </tr>
  );
};

export default UserRow;
