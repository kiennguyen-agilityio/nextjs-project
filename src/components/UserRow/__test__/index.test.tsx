import { render, screen } from '@testing-library/react';

// components
import UserRow from '@/components/UserRow';

// models
import { UserModel } from '@/models/UserModel';

jest.mock('@/api/role', () => ({
  getRoleById: jest.fn().mockResolvedValue({ id: 1, name: 'Admin' }),
}));

async function resolvedComponent(
  Component: (props: UserModel) => Promise<JSX.Element>,
  props: UserModel,
) {
  const ComponentResolved = await Component(props);
  return () => ComponentResolved;
}

describe('UserRow', () => {
  const mockProps: UserModel = {
    avatar: 'https://s.net.vn/Ur2Q',
    email: 'test@example.com',
    name: 'John Doe',
    userRole: 1,
    joined: new Date('2022-01-01'),
    id: '1',
  };

  it('should render all the content', async () => {
    const UserRowResolved = await resolvedComponent(UserRow, {
      ...mockProps,
    });
    const { asFragment } = render(
      <table>
        <tbody>
          <UserRowResolved />
        </tbody>
      </table>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders user information correctly', async () => {
    const UserRowResolved = await resolvedComponent(UserRow, {
      ...mockProps,
    });
    render(
      <table>
        <tbody>
          <UserRowResolved />
        </tbody>
      </table>,
    );

    expect(screen.getByAltText(mockProps.name)).toBeInTheDocument();
    expect(screen.getByText(mockProps.email)).toBeInTheDocument();
    expect(screen.getByText(mockProps.name)).toBeInTheDocument();
    expect(screen.getByText('Admin')).toBeInTheDocument();
    expect(screen.getByText('1/1/2022')).toBeInTheDocument();
  });

  it('renders the TripleDot component', async () => {
    const UserRowResolved = await resolvedComponent(UserRow, {
      ...mockProps,
    });
    render(
      <table>
        <tbody>
          <UserRowResolved />
        </tbody>
      </table>,
    );

    expect(screen.getByTestId('triple-dot')).toBeInTheDocument();
  });
});
