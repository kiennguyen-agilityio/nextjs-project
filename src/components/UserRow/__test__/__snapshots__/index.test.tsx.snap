// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserRow should render all the content 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr
        class="flex flex-col sm:flex-row items-start sm:items-center sm:gap-10 pr-8 md:pr-0 justify-between p-0 sm:p-4 border-b border-gray-200 hover:bg-gray-100 transition-colors duration-200 ease-in-out"
      >
        <td
          class="flex items-center w-full mb-4 sm:mb-0"
        >
          <img
            alt="John Doe"
            class="w-10 h-10 rounded-lg mr-4"
            data-nimg="1"
            decoding="async"
            height="40"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fs.net.vn%2FUr2Q&w=96&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fs.net.vn%2FUr2Q&w=48&q=75 1x, /_next/image?url=https%3A%2F%2Fs.net.vn%2FUr2Q&w=96&q=75 2x"
            style="color: transparent;"
            width="40"
          />
          <div
            class="text-sm font-medium sm:truncate max-w-24 xl:max-w-64 md:max-w-25"
          >
            test@example.com
          </div>
        </td>
        <td
          class="text-sm w-full truncate mb-4 sm:mb-0"
        >
          John Doe
        </td>
        <td
          class="text-sm w-full truncate mb-4 sm:mb-0"
        >
          Admin
        </td>
        <td
          class="text-sm w-full truncate mb-4 sm:mb-0"
        >
          1/1/2022
        </td>
        <td
          class="w-full sm:w-1/12"
        >
          <div
            class="relative"
          >
            <button
              aria-label="More options"
              class="text-gray-500 p-1 rounded-full hover:bg-gray-200"
              data-testid="triple-dot"
            >
              <svg
                class="w-6 h-6"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 12h.01M12 12h.01M19 12h.01"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="4"
                />
              </svg>
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
